CoroutineTemplate/data/src/main/java/co/nimblehq/coroutine/repositoryimpl/UserRepository.kt

```diff
@@ -0,0 +1,21 @@
+package co.nimblehq.coroutine.repositoryimpl
+
+import co.nimblehq.coroutine.data.service.ApiService
+import co.nimblehq.coroutine.data.service.response.toUserEntities
+import co.nimblehq.coroutine.entity.UserEntity
+import co.nimblehq.coroutine.repository.UserRepository
+import co.nimblehq.coroutine.usecase.UseCaseResult
+
+class UserRepositoryImpl constructor(
+    private val apiService: ApiService,
+) : UserRepository {
+
+    override suspend fun getUsers(): UseCaseResult<List<UserEntity>> {
+        return try {
+            val response = apiService.getUsers()
+            UseCaseResult.Success(response.toUserEntities())
+        } catch (e: Exception) {
+            UseCaseResult.Error(e)
```

Comment:
now I see the systematic issue here, @doannimble. Why would we let the `Repository` returns `UseCase` result? 🤔  it's kinda crossing the line. Imagine if you have a different Service that doesn't run through UseCase class, the `Repository` become very odd to returns `UseCaseResult` wrapper in such case.

Reply:
@sleepylee Perhaps the name of `UseCaseResult` makes you confuse.
It's just the `Data Model` of `Repository` and the `UseCase` must depend on it.
I usually use with name `Result`. Do I need to refactor the name of `UseCaseResult` to avoid confusion?

Reply:
I think it's better that we return the type as we did before for now. I see that there are some resources on Medium that refer to returns in this way, but it doesn't go along with our intention.

- Your approach is:
`Repository returns Result [ Success, Error] - UseCase is dumb, it returns the Result that belongs to Repository`

https://proandroiddev.com/why-you-need-use-cases-interactors-142e8a6fe576

- While our approach currently is:

`Repository returns raw <Type> and throw exception or onError with RxJava; UseCase has UseCaseResult.Success or Error that can be manipulated by the ViewModel`

https://github.com/android10/Android-CleanArchitecture/blob/master/domain/src/main/java/com/fernandocejas/android10/sample/domain/repository/UserRepository.java

Reply:
@sleepylee Okay, we may have another discussion about this in other PR.
I will keep our current approach and I think it is better if we have a deep understanding of each to answer the question: why we use this one, not that one.

Reply:
here is another ref I found that they followed your style (in this case, they returned `Resource` == your `Result`) at the Repository layer: https://github.com/android/architecture-components-samples/blob/88747993139224a4bb6dbe985adf652d557de621/GithubBrowserSample/app/src/main/java/com/android/example/github/repository/UserRepository.kt#L40

It's from Google's sample, notice the difference is that they are not using `Interactor` or `UseCase` like us, so their return type is being wrapped properly before sending back to the ViewModel directly.

For us, we segregated the responsibility more via an interim layer, so, as the thumb rule: we must return the wrapped type back to the ViewModel to avoid further processing; doing so at the UseCase can reduce the heavy logic at the Repository layer. Reducing the size of each class is always better than creating more dumb classes to invoke the other layer blindly imo. 🙏 

Let me know your thoughts in the end.

Reply:
In my humble opinion, wrapping the data inside `UseCaseResult` doesn't violate anything. But I think it would be better to leave this job to the UseCase. Sometimes the UseCase does manipulations on the data or has internal logic based on the data. In those cases, unwrapping it inside the UseCase seems redundant. So I would recommend making the method return `List<UserEntity>` only.

Reply:
@sleepylee @AVI5HEK Okay, so I agree with you guys, we should let `UseCase` handle all of the business logic, included wrapping data.
For example, let say if we want to save a token of `UserEntity` after logged in into` SharedPreference`, we can do it as a flow on try/catch inside `UseCase` without unwrapping data.